### PR TITLE
[wip] build: Switch from C++11 to C++17

### LINKED ISSUE #16684: Discussion: upgrading to C++17
**UPDATE**: Release plan summary from April 2nd, 2020 meeting:

|       | C++11 Compat       | C++17 Compat       | Gitian/Guix Release |
|-------|--------------------|--------------------|---------------------|
| v0.20 | :heavy_check_mark: | if easy            | C++11               |
| v0.21 | :heavy_check_mark: | :heavy_check_mark: | C++17               |
| v0.22 | :x:                | :heavy_check_mark: | C++17               |

Meeting logs can be found here: http://www.erisian.com.au/meetbot/bitcoin-core-dev/2020/bitcoin-core-dev.2020-04-02-19.01.log.html

-----

There were some discussion around upgrading to C++17. It seems to be able to give us library and language features that will simplify future code. ~~For example, for library features, our current `optional.h` is just a wrapper around `boost::optional`, and can be replaced with `std::optional`. Other library features that might be useful include (non-exhaustively): `std::variant`, the file system library, and `std::shared_mutex`.~~ @elichai has kindly provided a table summarizing features useful for Bitcoin Core here: https://github.com/bitcoin/bitcoin/issues/16684#issuecomment-524067463.

I want to apologize for misspeaking about C++17 _requiring_ GCC8, that is **untrue**, and I believe all of the features we would want will be covered by GCC7, and some by GCC6.

For some background on the **cons**: https://github.com/bitcoin/bitcoin/issues/13356#issuecomment-457870907

-----

Default GCC versions major distros (as of Aug 22nd):

|   | Operating System | release version | GCC version | Notes |
|---|------------------|--------------------------|-------------|------|
| :heavy_check_mark: | Debian           | Buster                   | 8.3         |      |
| :heavy_check_mark: | Ubuntu           | 18.04 LTS                | 7.4         |      |
| :heavy_check_mark: | Fedora           | 30                       | 9.1.1       |      |
| :heavy_check_mark: | RHEL             | 8                        | 8.2         |      |
|  | CentOS           | 7                        | 4.8.5       | CentOS 8 (GCC 8.2) is being released [as we speak](https://wiki.centos.org/About/Building_8#head-516d5e6556bb8523b52fba246953d32831582480) |

-----

In-depth code examples of

C++17 features: https://github.com/AnthonyCalandra/modern-cpp-features/blob/master/CPP17.md
C++14 features: https://github.com/AnthonyCalandra/modern-cpp-features/blob/master/CPP14.md

-----

Here are some compatibility tables from https://en.cppreference.com/w/cpp/compiler_support#cpp17

Library features:
Screen Shot 2019-08-22 at 19 39 24

Language features:
Screen Shot 2019-08-22 at 19 38 32

### PR DESCRIPTION
**Pull request by practicalswift**
Switch from C++11 to C++17.

Fixes #16684 ("Discussion: upgrading to C++17").

_**This is work in progress and thus not ready to merge. C++17 is optimistically planned for 0.21.0.**_

* build: Update `ax_cxx_compile_stdcxx.m4` to latest version which supports C++17 checking
* build: Use C++17 mode when compiling. Require a compiler with support for C++17 language features.
* ci: Add `-fsanitize=integer` suppression (`unsigned-integer-overflow`) for `libstdc++` C++17's `basic_string` (`rfind`)
* Use `std::make_unique` (C++14) instead of legacy `MakeUnique` wrapper
* Use `std::optional` (C++17) instead of `boost::optional`
* Use `[[nodiscard]]` (C++17) instead of legacy `NODISCARD` macro

Left to do:

- [ ] Make AppVeyor compile using MSVC in C++17 mode.
- [ ] Revert temporary commit "Temporarily disable Travis build jobs without a C++17 compiler" and apply proper fix instead
- [ ] Revert temporary commit "Temporarily disable `test_spanparsing`" (`test_spanparsing` does not not compile in C++17 mode) and apply proper fix instead

### CODE REVIEW

File: depends/packages/qt.mk [hunk 1]
```diff
@@ -23,7 +23,8 @@ define $(package)_set_vars
 $(package)_config_opts_release = -release
 $(package)_config_opts_debug = -debug
 $(package)_config_opts += -bindir $(build_prefix)/bin
-$(package)_config_opts += -c++std c++11
+# Note: C++17 not supported yet.
+$(package)_config_opts += -c++std c++14
```
Comment (emilengler) [inline depends/packages/qt.mk]: I'm not 100% sure on this but this answer says that it should work at Qt 5.12:
https://stackoverflow.com/questions/46610996/cant-use-c17-features-using-g-7-2-in-qtcreator

Not sure if all distros support Qt 5.12

File: build-aux/m4/ax_cxx_compile_stdcxx.m4 [hunk 2]
```diff
@@ -57,26 +59,13 @@ AC_DEFUN([AX_CXX_COMPILE_STDCXX], [dnl
...
         [$3], [optional], [ax_cxx_compile_cxx$1_required=false],
         [m4_fatal([invalid third argument `$3' to AX_CXX_COMPILE_STDCXX])])
-  m4_if([$4], [], [ax_cxx_compile_cxx$1_try_default=true],
-        [$4], [default], [ax_cxx_compile_cxx$1_try_default=true],
-        [$4], [nodefault], [ax_cxx_compile_cxx$1_try_default=false],
-        [m4_fatal([invalid fourth argument `$4' to AX_CXX_COMPILE_STDCXX])])
   AC_LANG_PUSH([C++])dnl
   ac_success=no
 
-  m4_if([$4], [nodefault], [], [dnl
```
Comment (fanquake) [inline build-aux/m4/ax_cxx_compile_stdcxx.m4]: You've dropped our changes here, was that intentional? If so, can you explain why they are no-longer required as part of the commit.

File: configure.ac [hunk 3]
```diff
@@ -61,8 +61,8 @@ case $host in
      lt_cv_deplibs_check_method="pass_all"
   ;;
 esac
-dnl Require C++11 compiler (no GNU extensions)
-AX_CXX_COMPILE_STDCXX([11], [noext], [mandatory], [nodefault])
+dnl Require C++17 compiler (no GNU extensions)
+AX_CXX_COMPILE_STDCXX([17], [noext], [mandatory], [nodefault])
```
Comment (fanquake) [inline configure.ac]: If dropping our `.m4` changes was intentional (see above), this call needs to be updated, as it no longer takes 4 arguments.

**Comment by elichai** [comment]
I'd say let's stick to build changes only here. code changes can come in a follow up PR

**Review by fanquake — COMMENTED** [review]
I think it's a bit too early to open this PR. Having a branch with the changes is interesting, and would make a good discussion point for #16684. 

However as a PR, it's going to be a long time until this is might be merged, and in the interim it's going to conflict with [a lot of other PRs](https://github.com/bitcoin/bitcoin/pull/17904#issuecomment-573146961), and generate a lot of noise.

I've left a couple comments inline where you've dropped our changes to the `ax_cxx_compile_stdcxx.m4` macro. Not sure if that was intentional.

**Comment by practicalswift** [comment]
@fanquake The `ax_cxx_compile_stdcxx.m4` was taken as-is from our upstream: I didn't know we had intentional deviations from upstream, so the reversion of those local changes was unintentional :)

Closing PR for now. Feel free to take on this as 0.21.0 is getting closer to merge - it is up for grabs :)

**Pull request closed without merging**